lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/participants/completion/MavenCompletionParticipant.java

```diff
@@ -860,4 +863,36 @@ private static CompletionItem toCompletionItem(String label, String description,
 		Map<Object, Boolean> map = new ConcurrentHashMap<>();
 		return t -> map.putIfAbsent(keyExtractor.apply(t), Boolean.TRUE) == null;
 	}
+	
+	private void internalCollectWorkspaceArtifacts(ICompletionRequest request, Collection<ArtifactInfo> artifactInfosCollector,
+			Optional<String> groupId, Optional<String> artifactId) {
+		DOMElement parent = request.getParentElement();
```

Comment:
I don't get why line 869..879 are necessary.

Reply:
We need to know what is the parent node for proper filtering
We need to know what is the grand parent in order to calculate the only when we're in `<parent/> `, `<dependency/>` and `<plugin/>` nodes, if we remove these lines the proposals will be calculated for any `<groupId/>`, `<artifactID/>` and `<version/>` node contents encountered anywhere iside the document, for example inside `<project/>` node.

Reply:
OK, but shouldn't this filter be happening in the caller method? This kind of filtering doesn't seem to be specific to workspace completion, they're also true for local repo and index completion, aren't they?

Reply:
There is actually only `<parent/>`/`<plugin/>` separation in the caller method to filter the proposals, not sure if we need to replace it with any kind of filtering that will include workspace proposals.
I believe workspace proposals are to be added into the list of proposals found in local/remote/etc. repository searchers, not replacing them.
So placing parent/dependency/plugin filter here (inside `internalCollectWorkspaceArtifacts` makes `onXMLContent` method to be a bit shorter by preventing the copying of this filter among the three different cases of the caller's switch.